modules/mod-deep-learning/DeepLearningEffect.cpp

```diff
@@ -0,0 +1,182 @@
+
+// register source separation
+namespace{ BuiltinEffectsModule::Registration< DeepLearningEffect > reg; }
+
+DeepLearningEffect::DeepLearningEffect()
+{
+   SetLinearEffectFlag(false);
+}
+
+DeepLearningEffect::~DeepLearningEffect()
+{
+}
+
+// ComponentInterface implementation
+
+ComponentInterfaceSymbol DeepLearningEffect::GetSymbol()
+{
+   return Symbol;
+}
+
+TranslatableString DeepLearningEffect::GetDescription()
+{
+   return XO("The goal of audio source separation is to isolate "
+             "the sound sources in a given mixture of sounds.");
+}
+
+ManualPageID DeepLearningEffect::ManualPage()
+{
+   return L"Source_Separation"; 
+}
+
+// EffectDefinitionInterface implementation
+
+EffectType DeepLearningEffect::GetType()
+{
+   return EffectTypeProcess;
+}
+
+// Effect implementation
+
+// ProcessOne() takes a track, transforms it to bunch of buffer-blocks,
+// performs a forward pass through the deep model, and writes 
+// the output to new tracks. 
+bool DeepLearningEffect::ProcessOne(WaveTrack *leader,
+                           double tStart, double tEnd)
+{
+   if (leader == NULL)
+      return false;
+
+   // keep track of the sample format and rate,
+   // we want to convert all output tracks to this
+   sampleFormat origFmt = leader->GetSampleFormat();
+   int origRate = leader->GetRate();
+
+   // initialize source tracks, one for each source that we will separate
+   std::vector<WaveTrack::Holder> sourceTracks;
+   std::vector<std::string> sourceLabels = mModel->GetCard()->labels();
+   sourceTracks = CreateSourceTracks(leader, sourceLabels);
+   
+   // Initiate processing buffer, most likely shorter than
+   // the length of the selection being processed.
+   Floats buffer{ leader->GetMaxBlockSize() };
+
+   // get each of the blocks we will process
+   std::vector<BlockIndex> indices = GetBlockIndices(leader, tStart, tEnd);
+   for (BlockIndex block : indices)
+   {
+      //Get a blockSize of samples (smaller than the size of the buffer)
+      sampleCount samplePos = block.first;
+      size_t blockSize = block.second;
+      double tPos = leader->LongSamplesToTime(samplePos); 
+   
+      // get a torch tensor from the leader track
+      torch::Tensor input = BuildMultichannelTensor(leader, buffer.get(), 
+                                            samplePos, blockSize);
+
+      // if we're not doing a multichannel forward pass, downmix
+      if (!mModel->GetCard()->multichannel())
+         input = input.sum(0, true, torch::kFloat);
+
+      // resample!
+      input = mModel->Resample(input, origRate, mModel->GetSampleRate());
+
+      // forward pass!
+      torch::Tensor output = mModel->ToTensor(ForwardPassInThread(input));
+
+      // resample back
+      output = mModel->Resample(output, mModel->GetSampleRate(), origRate);
+
+      // write each source output to the source tracks
+      for (size_t idx = 0; idx < output.size(0) ; idx++)
+         TensorToTrack(output[idx].unsqueeze(0), sourceTracks[idx], 
```

Comment:
Now here is another indexing into `sourceTracks`.  Whatever the first dimension of the output tensor is, update that many channels in sourceTracks.

So what guarantees that this output dimension is equal to (or, just not more than) the number of "labels"?

Reply:
this should work better
```cpp
for (size_t idx = 0; idx < std::min(sourceTracks.size(), (size_t)output.size(0)) ; idx++)
```